### ISSUE TITLE
Should Swift 2 support be folded into RAC 3.0?

### ISSUE THREAD

**Issue by jspahrsummers** [ReactiveCocoa/ReactiveCocoa#2187]
Now that RAC 3.0 is basically ready to go, should we fold in the Swift 2 support before shipping it? My sense is that most folks have migrated to Swift 2 already, but it _is_ still in beta.

If we _don't_ want to do that, Swift 2 support would constitute RAC 4.0. That isn't a big deal, really, but might be counterintuitive.

Thoughts?

/cc @ReactiveCocoa/reactivecocoa

**Comment by andymatuschak** [comment]
I’d suggest no: give the Swift 2 API the time and space it needs to evolve. A lot of clients are waiting on “an official release” of RAC 3; there’s no sense delaying them ‘till September (or likely later). And many clients won’t/can't update to Swift 2 for some time.

**Comment by jspahrsummers** [comment]
Good enough for me! I was leaning “no” anyways, just needed at least one +1 to that. :smile:

**Comment by JaviSoto** [comment]
>And many clients won’t/can't update to Swift 2 for some time.

I'm curious what reasons those might be, I'm probably not taking those into account.

While I totally agree with @andymatuschak, and I think that's a totally valid way to go forward, I'll play devil's advocate, cause I think shipping with Swift 2 support in RAC 3 may also be a good idea:

- It would only delay shipping RAC 3 "stable" (or rather, against a stable Swift release) for another 2 months.
- Like you said, a lot of people are probably writing Swift 2 code this summer.
- A lot of folks may have the impression that RAC is a very complex framework whose API tends to change a lot, and that may be discouraging. if RAC 4 brings changes that require tons of source changes, I'm afraid that may contribute to that.
- One of the least ideal things for people to learn RAC 3 may be the `|>` operator. Because it's a custom operator, because the method signature is not easy to grasp, but I think most importantly, because Xcode just fails at autocompleting methods that you can put on the right side (and if you use the wrong one, the type errors are very hard to understand). For this alone, I think the Swift 2 branch makes learning RAC in Swift a lot easier for newcomers, and shipping that directly would save them from having to learn `|>`.
- RAC 3 targeting Swift 1.2 would be stuck with an unmaintained release of Swift. It's going to be tough to get issues from people having compiler crashes / Swift runtime crashes that we can't do much about, and we know Apple won't fix (on 1.2).

**Comment by brow** [comment]
> > And many clients won’t/can't update to Swift 2 for some time.

> I'm curious what reasons those might be, I'm probably not taking those into account.

Swift 2 code can't be built except in Xcode 7, and iOS apps built in Xcode 7 cannot be submitted to the App Store, right?

> RAC 3 targeting Swift 1.2 would be stuck with an unmaintained release of Swift... For this alone, I think the Swift 2 branch makes learning RAC in Swift a lot easier for newcomers, and shipping that directly would save them from having to learn |>.

This point speaks to me, but the alternative for newcomers in the next few months is learning ReactiveCocoa 2 (and often learning to use it from Swift). I'm guiding a teammate through this right now, and I'm struggling to show him the value of the MVVM and signals underneath all the painful downcasting, constructing and deconstructing `RACTuple`s without macros, and liberal use of KVO through `@objc` and `dynamic` with unchecked keypaths for want of `PropertyType`.

**Comment by andymatuschak** [comment]
Ditto Tom's latter point there.

**Comment by iv-mexx** [comment]
> A lot of clients are waiting on “an official release” of RAC 3

Is it that big of a difference between having the current RC released as "official 3.0" release and just having the current RC? In earlier beta releases, it was clearly stated that these were beta releases and not ready for production yet.
Could it be enough to have a similar note in the RC ( _This RC is ready for production, but the final release of RAC 3.0 will target Swift 2_ )?

**Comment by jspahrsummers** [comment]
Honestly, this is a problem that probably won't be doing away in the foreseeable future—namely, it seems like _every_ Swift update is going to be breaking source compatibility, so we need a strategy in place for when that happens.

Might as well start here, and I'd rather not ship a stable release of RAC against an _unstable_ version of Swift.

**Comment by neilpa** [comment]
> Might as well start here, and I'd rather not ship a stable release of RAC against an unstable version of Swift.

This right here. RAC 3 has baked against Swift 1.2 for quite some time and it's not worth throwing that out at this point.

**Comment by JaviSoto** [comment]
All good points :+1: Thanks everyone!
